### ISSUE TITLE
Sipi and subfolders

### ISSUE THREAD

**Issue by subotic** [dasch-swiss/sipi#288]
It is not practical to use subfolders with Sipi as it needs to be restarted when files are added.

At some point in time, we will have a decent number of projects, and shutting down Sipi, just to add a few files through copying, will not be feasible.

We have two possibilities:
 1. We look at the folder structure of Sipi as a closed black box and only interact with Sipi through a management API and a command line tool, which allows to add and retrieve files and make backups
 2. Sipi is vigilant and doesn't care about the subfolder structure. Whatever is there, Sipi analyses and finds the files to serve.

The current automatic generation of subfolders is fine for case 1, but for case 2 the folder structure should be freely definable.

To sum it up, it is a pain to work the subfolders. Looking at https://stackoverflow.com/questions/466521/how-many-files-can-i-put-in-a-directory/466596#466596 for ext4 this is a non-issue.

**Comment by benjamingeer** [comment]
Lukas wants to implement a Sipi route to add a new project without shutting down Sipi:

https://github.com/dhlab-basel/Sipi/issues/281#issuecomment-461419894

**Comment by benjamingeer** [comment]
The StackOverflow answer you cited only gives the theoretical maximum number of files for directory. It doesn't say anything about performance. This answer may be more useful:

https://stackoverflow.com/a/11373532

**Comment by loicjaouen** [comment]
@subotic  "for ext4 this is a non-issue" : i didn't get what the issue was?

We use subfolders for:
- not being hit by a limit of files in subfolder : this is not an issue any more
- file access time being slow on large folders : this is not an issue any more (!)
- for not putting all eggs in one basket : being able to recover if sipi (as any other tool) breaks down

I think that sipi should create the projects subfolders on the fly.

For day-to-day managements, for now, I find useful that images are set in different folders (it is easier to clean-up a failed import).

**Comment by benjamingeer** [comment]
"Either of these filesystems can handle tens of thousands of files in a single directory, with XFS being significantly faster than ext4 on a directory with the same number of inodes. But you probably don't want a single directory with 3M inodes"

https://serverfault.com/a/796696

It would be nice to have some real data on this (and also to know which filesystem SWITCHengines is using).

**Comment by benjamingeer** [comment]
@loicjaouen I think @subotic is talking about the directories that Sipi creates below each project folder. For example, if you have a project `0803`, Sipi creates `0803/A`, `0803/B`, etc.

**Comment by loicjaouen** [comment]
ok... subfolders are [A..Z], which are created by sipi at start-up.
So the test case is: if I could to copy some files by hand, I create the subfolders and copy the files but sipi won't see them? 

Regarding performances, any recent file system will use a tree hashed index with "fast" access to the files... if coded correctly, which might not be the case for some tools (even `ls`) or some usages (with 1M files it just doesn't make sense to list them).

**Comment by lrosenth** [comment]
SWITCH uses CEPH - I do not know how CEPH does index the files.  As far as I know all modern file systems are using some kind of tree indexing - thus 1 Mio files should not pose a real problem. I’m very ambiguous about using SIPI with “subdirs”. On one hand it adds a complication layer, on the other hand it reduces the number of files per directory (which might or might not be necessary...).

I’ll do some research about it to have a better decision base...

Lukas

Sent from my iPad

On 29 Mar 2019, at 10:00, Loïc Jaouen <notifications@github.com<mailto:notifications@github.com>> wrote:


ok... subfolders are [A..Z], which are created by sipi at start-up.
So the test case is: if I could to copy some files by hand, I create the subfolders and copy the files but sipi won't see them?

Regarding performances, any recent file system will use a tree hashed index with "fast" access to the files... if coded correctly, which might not be the case for some tools (even ls) or some usages (with 1M files it just doesn't make sense to list them).

**Comment by lrosenth** [comment]
Since using subdirs is optional (by seting dirleves to "0", no subduers are used, I leave it in the code. There may be applications where it can make sesnse..
